src/mlpack/methods/decision_tree/all_categorical_split.hpp

```diff
@@ -47,45 +51,80 @@ class AllCategoricalSplit
    *      splitting.
-   * @param classProbabilities Class probabilities vector, which may be filled
-   *      with split information a successful split.
+   * @param splitInfo Stores split information on a successful split.
    * @param minimumGainSplit Minimum  gain split.
    * @param aux Auxiliary split information, which may be modified on a
    *      successful split.
    */
-  template<bool UseWeights, typename VecType, typename WeightVecType>
+  template<bool UseWeights, typename VecType, typename LabelsType,
+           typename WeightVecType>
   static double SplitIfBetter(
       const double bestGain,
       const VecType& data,
       const size_t numCategories,
-      const arma::Row<size_t>& labels,
+      const LabelsType& labels,
       const size_t numClasses,
       const WeightVecType& weights,
       const size_t minimumLeafSize,
       const double minimumGainSplit,
-      arma::vec& classProbabilities,
+      arma::vec& splitInfo,
+      AuxiliarySplitInfo& aux);
+
+  /**
+   * Check if we can split a node.  If we can split a node in a way that
+   * improves on 'bestGain', then we return the improved gain.  Otherwise we
+   * return the value 'bestGain'.  If a split is made, then splitInfo and
+   * aux may be modified.  For this particular split type, aux will be empty
+   * and splitInfo will store the number of children of the node.
+   *
+   * This overload is used only for regression.
+   *
+   * @param bestGain Best gain seen so far (we'll only split if we find gain
+   *      better than this).
+   * @param data The dimension of data points to check for a split in.
+   * @param numCategories Number of categories in the categorical data.
+   * @param responses Responses for each point.
+   * @param weights Weights associated with responses.
+   * @param minimumLeafSize Minimum number of points in a leaf node for
+   *      splitting.
+   * @param splitInfo Stores split information on a successful split.
+   * @param minimumGainSplit Minimum  gain split.
+   * @param aux Auxiliary split information, which may be modified on a
+   *      successful split.
+   */
+  template<bool UseWeights, typename VecType, typename ResponsesType,
+           typename WeightVecType>
+  static double SplitIfBetter(
+      const double bestGain,
+      const VecType& data,
+      const size_t numCategories,
+      const ResponsesType& responses,
+      const WeightVecType& weights,
+      const size_t minimumLeafSize,
+      const double minimumGainSplit,
+      double& splitInfo,
       AuxiliarySplitInfo& aux);
 
   /**
    * Return the number of children in the split.
    *
-   * @param classProbabilities Auxiliary information for the split.
+   * @param splitInfo Auxiliary information for the split.
    * @param * (aux) Auxiliary information for the split (Unused).
    */
-  static size_t NumChildren(const arma::vec& classProbabilities,
+  static size_t NumChildren(const double& splitInfo,
```

Comment:
Yeah. We can do that. From the source code point of view, this will be a great step because it will greatly simplify the implementation without any of these clever techniques like empty struct, reusing variables etc. :)

Reply:
Since we are talking about optimization here, it came to my mind today that can't we parallelize the gain computation for each feature while calculating the best dimension?🤔

Reply:
We talked about this last week, but just to write down my thoughts, I do think that it could be possible to use OpenMP parallel for loops in `DecisionTree<>::Train()`... I think that will work okay.  If we try it, we should make sure to test the random forest runtime (which is parallelized at the level of trees), to make sure that we don't have any slowdown or anything.